docs-user/bunny.md

```diff
@@ -0,0 +1,272 @@
+# Case Study
+## 2D canvas and worker messaging
+
+The following article is a case study in using the profiler to identify a performance issue, and then fixing it for a 4 times performance improvement. This improvement moves the frame rate from a fairly slow 15fps to a smooth 60fps.
+
+## The project description
+
+![A picture of a 3d bunny rabbit model rendered using small squares.](./images/bunny-analysis/bunny.png)
+
+The project is a website that takes user's JavaScript code, and runs it to produce a visualization. The user's code only has access to the function `rect(color, x, y, width, height)`, which draws a rectangle to the screen. For the implementation, the website posts the user's code to a sandboxed iframe. The iframe has a `<canvas>` element that is rendered to via the [`CanvasRenderingContext2D`](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D) API. The user's code is evaluated in a WebWorker, and then the results of `rect()` are posted back to the iframe's code.
+
+In a simplified example, the `worker.js` would run something like:
+
+```js
+// Evaluate the user's code string:
+eval(`
+rect('#fff', 10, 10, 3, 3);
+rect('#fff', 20, 20, 3, 3);
+rect('#fff', 30, 30, 3, 3);
+rect('#fff', 40, 40, 3, 3);
+rect('#fff', 50, 50, 3, 3);
+`);
+
+// worker.js would post something like this to the iframe:
+self.postMessage({
+  color: ['#fff', '#fff', '#fff', '#fff', '#fff'],
+  x: [10, 20, 30, 40, 50],
+  y: [10, 20, 30, 40, 50],
+  w: [3, 3, 3, 3, 3],
+  h: [3, 3, 3, 3, 3],
+});
+```
+
+Then the iframe would draw the code using:
+
+```js
+worker.addEventListener('message', message => {
+  const { data } = message;
+  for (let i = 0; i < data.color.length; i++) {
+    ctx.fillStyle = data.color[i];
+    ctx.fillRect(data.x[i], data.y[i], data.w[i], data.h[i]);
+  }
+});
+```
+
+This would result in user's evaluated code drawing something to the screen, like the bunny model in the above image.
+
+## The problem
+
+This code ended up not scaling well for large sets of rectangles being drawn to the screen. There were lots of stutters, and a slow frame rate. The user impact for fixing this problem would be to have a smoother frame rate, plus the ability to draw many more rectangles to the screen without slowing things down. In order to validate fixes, the following steps were used to reproduce the issue.
+
```

Comment:
I think that before explaining the exact process to get a profile, I'd explain again the general logic to fix a performance issue, that is:
1. Measure
2. Fix
3. Measure again

And maybe also structure the page with these parts.

Reply:
I added a section to the opening paragraph about this.